pkg/apis/rollouts/validation/validation_references.go

```diff
@@ -214,25 +214,45 @@ func ValidateIngress(rollout *v1alpha1.Rollout, ingress v1beta1.Ingress) field.E
 }
 
 func ValidateVirtualService(rollout *v1alpha1.Rollout, obj unstructured.Unstructured) field.ErrorList {
-	allErrs := field.ErrorList{}
+	var fldPath *field.Path
+	var virtualServices []v1alpha1.IstioVirtualService
 	newObj := obj.DeepCopy()
-	fldPath := field.NewPath("spec", "strategy", "canary", "trafficRouting", "istio", "virtualService", "name")
-	vsvcName := rollout.Spec.Strategy.Canary.TrafficRouting.Istio.VirtualService.Name
-	httpRoutesI, err := istio.GetHttpRoutesI(newObj)
-	if err != nil {
-		msg := fmt.Sprintf("Unable to get HTTP routes for Istio VirtualService")
-		allErrs = append(allErrs, field.Invalid(fldPath, vsvcName, msg))
-	}
-	httpRoutes, err := istio.GetHttpRoutes(newObj, httpRoutesI)
-	if err != nil {
-		msg := fmt.Sprintf("Unable to get HTTP routes for Istio VirtualService")
-		allErrs = append(allErrs, field.Invalid(fldPath, vsvcName, msg))
+	allErrs := field.ErrorList{}
+
+	if istioutil.MultipleVirtualServiceConfigured(rollout) {
+		fldPath = field.NewPath("spec", "strategy", "canary", "trafficRouting", "istio", "virtualServices", "name")
+		virtualServices = rollout.Spec.Strategy.Canary.TrafficRouting.Istio.VirtualServices
+	} else {
+		fldPath = field.NewPath("spec", "strategy", "canary", "trafficRouting", "istio", "virtualService", "name")
+		virtualServices = []v1alpha1.IstioVirtualService{rollout.Spec.Strategy.Canary.TrafficRouting.Istio.VirtualService}
 	}
-	err = istio.ValidateHTTPRoutes(rollout, httpRoutes)
-	if err != nil {
-		msg := fmt.Sprintf("Istio VirtualService has invalid HTTP routes. Error: %s", err.Error())
-		allErrs = append(allErrs, field.Invalid(fldPath, vsvcName, msg))
+
```

Comment:
Since the PR makes both VirtualService and VirtualServices as optional, we need to check if virtualservice is empty the validation. Therefore, I think the new function `ValidateRolloutReferencedResources` needs to be called here to make it consistent with the validation logic. WDYT?

Reply:
This is already taken care here 
https://github.com/argoproj/argo-rollouts/pull/1381/files#diff-7fecb947ba065e9c2f022d05ca775a0655d2c94e1776b89b95291fed79924df9R177
this is called by GetReferencedVirtualServices method which is called before the validation.

Reply:
@darshanhs09, what I meant is that the new validation function validateRolloutVirtualServicesConfig should be called before GetReferencedVirtualServices. validateRolloutVirtualServicesConfig should be condolidated with existing virtual service check at

https://github.com/argoproj/argo-rollouts/blob/2620490c9ab96e9da43b9c014fdf802c5e87dfa3/pkg/apis/rollouts/validation/validation_references.go#L216

which is done before reconcile.

Reply:
@huikang 
Please check the below code. getRolloutReferencedResources  is called before ValidateRolloutReferencedResources.
As i mentioned, the reason for introducing this validation in GetReferencedVirtualServices is because this method iterates the virtualservices to fill the k8s virtualservice record references. So, it doesn't really make sense to introduce this validation in ValidateVirtualService. But i can do one thing if required, before calling the method GetReferencedVirtualServices i can validate the multi VSes invalid config.
[https://github.com/argoproj/argo-rollouts/blob/2620490c9ab96e9da43b9c014fdf802c5e87dfa3/rollout/controller.go#L468](url)

Please let  me know your thoughts

Reply:
Hi, @darshanhs09, 

> But i can do one thing if required, before calling the method GetReferencedVirtualServices i can validate the multi VSes invalid config.

This is exactly what I suggested in my last comment (https://github.com/argoproj/argo-rollouts/pull/1381#discussion_r682662879), right? Please do so. Thanks.

Reply:
@huikang not really! It doesn't really make sense to validate (validateRolloutVirtualServicesConfig) in ValidateVirtualService because it is called after getRolloutReferencedResources. As i mentioned, i am validating before GetReferencedVirtualServices. Please check the latest update.